common/src/stack/pylib/stack/firmware.py

```diff
@@ -0,0 +1,98 @@
+# Require the supported hash algorithms to be the always present ones
+SUPPORTED_HASH_ALGS = hashlib.algorithms_guaranteed
+
+class FirmwareError(Exception):
+	"""The exception type raised by the firmware utilities in this module."""
+	pass
+
+def calculate_hash(file_path, hash_alg, hash_value = None, digest_length = 256):
+	"""Calculates the hash of the provided file using the provided algorithm and returns it as a hex string.
+
+	hash_alg is required to be one of the SUPPORTED_HASH_ALGS and a FirmwareError will be raised if it is not.
+
+	If a hash value is provided, this checks the calculated hash against the provided hash. The hash_value should
+	be a string of the form provided by hash.hexdigest(). A FirmwareError is raised if the hashes do not match.
+
+	For some algorithms a length is required (shake_128 and shake_256 at the time of this writing). The digest_length
+	parameter allows the overriding of the default length used. This parameter is ignored if the algorithm does not
+	allow specifying the digest length.
+	"""
+	if hash_alg not in SUPPORTED_HASH_ALGS:
+		raise FirmwareError(
+			f"hash_alg must be one of the following: {SUPPORTED_HASH_ALGS}"
+		)
+
+	hasher = hashlib.new(name = hash_alg, data = Path(file_path).read_bytes())
+	try:
+		calculated_hash = hasher.hexdigest()
+	except TypeError:
+		calculated_hash = hasher.hexdigest(digest_length)
+
+	# check the hash if one was provided to check against
+	if hash_value is not None and hash_value != calculated_hash:
+		raise FirmwareError(
+			f"Calculated hash {calculated_hash} does not match expected hash {hash_value}. Algorithm was {hash_alg}."
+		)
+
+	return calculated_hash
+
+def fetch_firmware(source, make, model, filename = None, **kwargs):
+	"""Fetches the firmware file from the provided source and copies it into a stacki managed file.
+
+	source should be the URL from which to pull the firmware image from. If this is not one of the
+	supported schemes, a FirmwareError is raised.
+
+	make and model must be set to the make and model the firmware image applies to.
+
+	filename is the optional file name to write the image out to locally. This does not change the
+	destination folder, only the file name.
+
+	The remaining kwargs will be captured and passed through as necessary to the underlying mechanism
+	used to fetch the file from the source. For example, fetching via HTTP might need a username and
+	a password for authentication.
+
+	A FirmwareError is raised if fetching the file from the source fails.
+	"""
+	# parse the URL to figure out how we're going to fetch it
+	url = urlparse(url = source)
+
+	# build file path to write out to
+	dest_dir = BASE_PATH / make / model
+	dest_dir = dest_dir.resolve()
+	dest_dir.mkdir(parents = True, exist_ok = True)
+	# set a random file name if the name is not set
+	final_file = dest_dir / (uuid.uuid4().hex if filename is None else filename)
+
+	# copy from local file
+	if url.scheme == SUPPORTED_SCHEMES[0]:
```

Comment:
Enforcing via error / codifying SUPPORTED_SCHEMES is good, but I think this code is less readable than `if url.scheme == 'file':`, and then you don't need the comment anymore.

Reply:
I'll have to revisit this. I think I was using the variable elsewhere for error messages or something. Otherwise I would have done it inline.

Reply:
You are using it for the error messages below, which I think makes sense; I just don't think indexing into it here is good.

Reply:
Yeah, I agree. Maybe namedtuple something something. Let me look into it.

Reply:
How about now @bsanders?